azure-kusto-ingest/src/resource_manager/resource_uri.rs

```diff
@@ -0,0 +1,224 @@
+use azure_core::ClientOptions;
+use azure_storage::StorageCredentials;
+use azure_storage_blobs::prelude::{ClientBuilder, ContainerClient};
+use azure_storage_queues::{QueueClient, QueueServiceClientBuilder};
+use url::Url;
+
+#[derive(Debug, thiserror::Error)]
+pub enum ResourceUriError {
+    #[error("URI scheme must be 'https', was '{0}'")]
+    InvalidScheme(String),
+
+    #[error("Object name is missing in the URI")]
+    MissingObjectName,
+
+    #[error("SAS token is missing in the URI as a query parameter")]
+    MissingSasToken,
+
+    #[error("Account name is missing in the URI")]
+    MissingAccountName,
+
+    #[error(transparent)]
+    ParseError(#[from] url::ParseError),
+
+    #[error(transparent)]
+    AzureError(#[from] azure_core::Error),
+}
+
+/// Parsing logic of resource URIs as returned by the Kusto management endpoint
+#[derive(Debug, Clone)]
+pub(crate) struct ResourceUri {
+    pub(crate) service_uri: String,
+    pub(crate) object_name: String,
+    pub(crate) account_name: String,
+    pub(crate) sas_token: StorageCredentials,
+}
+
+impl TryFrom<&str> for ResourceUri {
+    type Error = ResourceUriError;
+
+    fn try_from(uri: &str) -> Result<Self, Self::Error> {
+        let parsed_uri = Url::parse(uri)?;
+
+        let scheme = match parsed_uri.scheme() {
+            "https" => "https".to_string(),
+            other_scheme => return Err(ResourceUriError::InvalidScheme(other_scheme.to_string())),
+        };
+
+        let host_string = parsed_uri
+            .host_str()
+            .expect("Url::parse should always return a host for a URI");
```

Comment:
Not a big fan of the panic here.
Use the host() function, also make sure it's not an Ip address (the Host enum has a Domain variant), since then you can't take the component.